### PR TITLE
[bazel, doc] Initial documentation for the new build & test rules

### PR DESCRIPTION
**Pull request by cfrantz**

### CODE REVIEW

File: rules/opentitan/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+
+## The `opentitan_test` rule
+
+The `opentitan_test` rule is actually a macro exported by `defs.bzl`.
+There is a real `_opentitan_test` rule in `cc.bzl`, but users are expected to use the macro.
+The reason for using a macro is to instantiate an individual test rule per exec\_env so that tests can be filtered by rule name or test tags.
+
+Like the binary rule, the test rule builds and links a binary, but then dispatches that binary according to the exec\_env-specific dispatch function and
+the test parameters in the `exec_env`.
```
Comment (jwnrt) [inline rules/opentitan/README.md:68]: Nit:

```suggestion
Like the binary rule, the test rule builds and links a binary, but then dispatches that binary according to the exec\_env-specific dispatch function and the test parameters in the `exec_env`.
```

**Review by jwnrt — APPROVED** [review]
LGTM, thanks @cfrantz 

Would you consider adding an entry to `SUMMARY.md` so we can link people to this doc on the website? Perhaps under the **Software** heading?

File: rules/opentitan/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  - An exec\_env-specific provider, such as `Cw310BinaryInfo` or `SimVerilatorBinaryInfo`.
+  - An exec\_env-specific `transform` function that converts binaries into the environment's required form (e.g. `vmem` files for DV or Verilator).
+- Testing
+  - An exec\_env-specific `test_dispatch` function that configures the test environment and dispatches the binary to that environment.
+
+The exec\_envs are provided to the `opentitan_binary` and `opentitan_test` rules in order to allow them to perform their function.
+
+## The `opentitan_binary` rule
+
+The `opentitan_binary` rule is very similar to bazel's built-in `cc_binary` rule: it accepts a list of sources, dependencies and other compiler configuration parameters and produces linked binary.
```
Comment (timothytrippel) [inline rules/opentitan/README.md:27]:
```suggestion
The `opentitan_binary` rule is very similar to bazel's built-in `cc_binary` rule: it accepts a list of sources, dependencies and other compiler configuration parameters and produces linked a binary.
```

**Review by timothytrippel — APPROVED** [review]
Eventually, it may be nice to link this to this page: https://opentitan.org/book/doc/contributing/bazel_notes.html

File: rules/opentitan/README.md (hunk 2 above)
Comment (jwnrt) [reply rules/opentitan/README.md:27]: Double suggestion!

```suggestion
The `opentitan_binary` rule is very similar to bazel's built-in `cc_binary` rule: it accepts a list of sources, dependencies and other compiler configuration parameters and produces a linked binary.
```

**Review by pamaury — APPROVED** [review]

File: rules/opentitan/README.md (hunk 1 above)
Comment (cfrantz) [reply rules/opentitan/README.md:68]: Fixed.

File: rules/opentitan/README.md (hunk 2 above)
Comment (cfrantz) [reply rules/opentitan/README.md:27]: Fixed.

**Comment by cfrantz** [comment]
> Would you consider adding an entry to `SUMMARY.md` so we can link people to this doc on the website? Perhaps under the **Software** heading?

Done:  Added a "Build & Test Rules" item under **Software**.

**Pull request merged**